File: objective-c-sample-v2/SendBirdSampleUI/SendBird/SendBirdMessaging/View/MessagingFileLinkTableViewCell.m
//
//  MessagingFileLinkTableViewCell.m
//  SendBirdSampleUI
//
//  Created by SendBird Developers on 2015. 7. 29..
//  Copyright (c) 2015 SENDBIRD.COM. All rights reserved.
//

#import "MessagingFileLinkTableViewCell.h"

#define kFileLinkCellTopMargin 14
#define kFileLinkCellBottomMargin 0
#define kFileLinkCellLeftMargin 12
#define kFileLinkFontSize 14.0
#define kFileLinkBalloonTopPadding 12
#define kFileLinkBalloonBottomPadding 12
#define kFileLinkBalloonLeftPadding 60
#define kFileLinkBalloonRightPadding 12
#define kFileLinkWidth 150
#define kFileLinkHeight 150
#define kFileLinkProfileHeight 36
#define kFileLinkProfileWidth 36
#define kFileLinkDateTimeLeftMarign 4
#define kFileLinkDateTimeFontSize 10.0
#define kFileLinkNicknameFontSize 12.0

@implementation MessagingFileLinkTableViewCell

- (id)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    
    if (self != nil) {
        [self initViews];
    }
    
    return self;
}

- (void) initViews
{
    [self setBackgroundColor:[UIColor clearColor]];
    
    self.profileImageView = [[UIImageView alloc] init];
    [self.profileImageView setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.profileImageView.layer setCornerRadius:(kFileLinkProfileHeight / 2)];
    [self.profileImageView setClipsToBounds:YES];
    [self.profileImageView setContentMode:UIViewContentModeScaleAspectFill];
    [self addSubview:self.profileImageView];
    
    self.messageBackgroundImageView = [[UIImageView alloc] init];
    [self.messageBackgroundImageView setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.messageBackgroundImageView setImage:[UIImage imageNamed:@"_bg_chat_bubble_gray"]];
    [self addSubview:self.messageBackgroundImageView];
    
    self.nicknameLabel = [[UILabel alloc] init];
    [self.nicknameLabel setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.nicknameLabel setFont:[UIFont systemFontOfSize:kFileLinkNicknameFontSize]];
    [self.nicknameLabel setNumberOfLines:1];
    [self.nicknameLabel setTextColor:UIColorFromRGB(0xa792e5)];
    [self.nicknameLabel setLineBreakMode:NSLineBreakByCharWrapping];
    [self addSubview:self.nicknameLabel];
    
    self.dateTimeLabel = [[UILabel alloc] init];
    [self.dateTimeLabel setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.dateTimeLabel setNumberOfLines:1];
    [self.dateTimeLabel setTextColor:UIColorFromRGB(0xacaab2)];
    [self.dateTimeLabel setFont:[UIFont systemFontOfSize:kFileLinkDateTimeFontSize]];
    [self.dateTimeLabel setText:@"11:24 PM"];
    [self addSubview:self.dateTimeLabel];
    
    self.fileImageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"_icon_file"]];
    [self.fileImageView setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.fileImageView setContentMode:UIViewContentModeScaleAspectFit];
    [self addSubview:self.fileImageView];
    
    // Profile Image View
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.profileImageView
                                                     attribute:NSLayoutAttributeBottom
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeBottom
                                                    multiplier:1 constant:-kFileLinkCellBottomMargin]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.profileImageView
                                                     attribute:NSLayoutAttributeLeading
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeLeading
                                                    multiplier:1 constant:kFileLinkCellLeftMargin]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.profileImageView
                                                     attribute:NSLayoutAttributeWidth
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:kFileLinkProfileWidth]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.profileImageView
                                                     attribute:NSLayoutAttributeHeight
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:kFileLinkProfileHeight]];
    
    // Nickname Label
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.nicknameLabel
                                                     attribute:NSLayoutAttributeBottom
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.fileImageView
                                                     attribute:NSLayoutAttributeTop
                                                    multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.nicknameLabel
                                                     attribute:NSLayoutAttributeLeading
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeLeading
                                                    multiplier:1 constant:kFileLinkBalloonLeftPadding]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.nicknameLabel
                                                     attribute:NSLayoutAttributeWidth
                                                     relatedBy:NSLayoutRelationLessThanOrEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:kFileLinkWidth]];
    
    // File Image View
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.fileImageView
                                                     attribute:NSLayoutAttributeBottom
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.messageBackgroundImageView
                                                     attribute:NSLayoutAttributeBottom
                                                    multiplier:1 constant:-kFileLinkBalloonBottomPadding]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.fileImageView
                                                     attribute:NSLayoutAttributeLeading
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeLeading
                                                    multiplier:1 constant:kFileLinkBalloonLeftPadding]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.fileImageView
                                                     attribute:NSLayoutAttributeWidth
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:kFileLinkWidth]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.fileImageView
                                                     attribute:NSLayoutAttributeHeight
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:kFileLinkHeight]];
    
    // Message Background Image View
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.messageBackgroundImageView
                                                     attribute:NSLayoutAttributeBottom
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeBottom
                                                    multiplier:1 constant:kFileLinkCellBottomMargin]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.messageBackgroundImageView
                                                     attribute:NSLayoutAttributeLeading
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeLeading
                                                    multiplier:1 constant:kFileLinkBalloonLeftPadding - 16]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.messageBackgroundImageView
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.fileImageView
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:kFileLinkBalloonRightPadding]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.messageBackgroundImageView
                                                     attribute:NSLayoutAttributeTop
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.nicknameLabel
                                                     attribute:NSLayoutAttributeTop
                                                    multiplier:1 constant:-kFileLinkBalloonTopPadding]];
    //    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.messageBackgroundImageView
    //                                                     attribute:NSLayoutAttributeTop
    //                                                     relatedBy:NSLayoutRelationEqual
    //                                                        toItem:self
    //                                                     attribute:NSLayoutAttributeTop
    //                                                    multiplier:1 constant:kFileLinkCellTopMargin]];
    
    // DateTime Label
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.dateTimeLabel
                                                     attribute:NSLayoutAttributeBottom
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.messageBackgroundImageView
                                                     attribute:NSLayoutAttributeBottom
                                                    multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.dateTimeLabel
                                                     attribute:NSLayoutAttributeLeading
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.messageBackgroundImageView
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:kFileLinkDateTimeLeftMarign]];
}

- (void) setModel:(SendBirdFileLink *)model
{
    self.fileLink = model;
    SendBirdSender *sender = [self.fileLink sender];
    long long ts = [self.fileLink getMessageTimestamp] / 1000;
    [self.dateTimeLabel setText:[SendBirdUtils messageDateTime:ts]];
    [self.nicknameLabel setText:[sender name]];
    
    [SendBirdUtils loadImage:[sender imageUrl] imageView:self.profileImageView width:kFileLinkProfileWidth height:kFileLinkProfileWidth];

    
    if ([[[self.fileLink fileInfo] type] hasPrefix:@"image"]) {
        [SendBirdUtils loadImage:[[model fileInfo] url] imageView:self.fileImageView width:kFileLinkWidth height:kFileLinkHeight];
    }
}

- (CGFloat)getHeightOfViewCell:(CGFloat)totalWidth
{
    NSString *nickname = [[self.fileLink sender] name];
    CGRect nicknameRect;
    NSMutableDictionary *nicknameAttribute = [NSMutableDictionary dictionaryWithObjectsAndKeys:[UIFont systemFontOfSize:12], NSFontAttributeName, nil];
    NSAttributedString *attributedNickname = [[NSAttributedString alloc] initWithString:nickname attributes:nicknameAttribute];
    
    nicknameRect = [attributedNickname boundingRectWithSize:CGSizeMake(kFileLinkWidth, CGFLOAT_MAX) options:(NSStringDrawingUsesLineFragmentOrigin) context:nil];
    
    return nicknameRect.size.height + kFileLinkCellBottomMargin + kFileLinkBalloonBottomPadding + kFileLinkHeight + kFileLinkBalloonTopPadding + kFileLinkCellTopMargin;
}

@end

File: swift-sample/SendBirdSampleUI/SendBird/SendBirdMessaging/View/MessagingFileLinkTableViewCell.swift
//
//  MessagingFileLinkTableViewCell.swift
//  SendBirdSampleUI
//
//  Created by Jed Kyung on 2/7/16.
//  Copyright © 2016 SENDBIRD.COM. All rights reserved.
//

import UIKit
import SendBirdSDK

class MessagingFileLinkTableViewCell: UITableViewCell {
    let kFileLinkCellTopMargin: CGFloat = 14
    let kFileLinkCellBottomMargin: CGFloat = 0
    let kFileLinkCellLeftMargin: CGFloat = 12
    let kFileLinkFontSize: CGFloat = 14.0
    let kFileLinkBalloonTopPadding: CGFloat = 12
    let kFileLinkBalloonBottomPadding: CGFloat = 12
    let kFileLinkBalloonLeftPadding: CGFloat = 60
    let kFileLinkBalloonRightPadding: CGFloat = 12
    let kFileLinkWidth: CGFloat = 150
    let kFileLinkHeight: CGFloat = 150
    let kFileLinkProfileHeight: CGFloat = 36
    let kFileLinkProfileWidth: CGFloat = 36
    let kFileLinkDateTimeLeftMarign: CGFloat = 4
    let kFileLinkDateTimeFontSize: CGFloat = 10.0
    let kFileLinkNicknameFontSize: CGFloat = 12.0

    var fileLink: SendBirdFileLink?
    var profileImageView: UIImageView?
    var fileImageView: UIImageView?
    var messageBackgroundImageView: UIImageView?
    var nicknameLabel: UILabel?
    var dateTimeLabel: UILabel?
    
    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        self.initViews()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func initViews() {
        self.backgroundColor = UIColor.clearColor()
        
        self.profileImageView = UIImageView()
        self.profileImageView?.translatesAutoresizingMaskIntoConstraints = false
        self.profileImageView?.layer.cornerRadius = kFileLinkProfileHeight / 2
        self.profileImageView?.clipsToBounds = true
        self.profileImageView?.contentMode = UIViewContentMode.ScaleAspectFill
        self.addSubview(self.profileImageView!)
        
        self.messageBackgroundImageView = UIImageView()
        self.messageBackgroundImageView?.translatesAutoresizingMaskIntoConstraints = false
        self.messageBackgroundImageView?.image = UIImage.init(named: "_bg_chat_bubble_gray")
        self.addSubview(self.messageBackgroundImageView!)
        
        self.nicknameLabel = UILabel()
        self.nicknameLabel?.translatesAutoresizingMaskIntoConstraints = false
        self.nicknameLabel?.font = UIFont.systemFontOfSize(kFileLinkNicknameFontSize)
        self.nicknameLabel?.numberOfLines = 1
        self.nicknameLabel?.textColor = SendBirdUtils.UIColorFromRGB(0xa792e5)
        self.nicknameLabel?.lineBreakMode = NSLineBreakMode.ByCharWrapping
        self.addSubview(self.nicknameLabel!)
        
        self.dateTimeLabel = UILabel()
        self.dateTimeLabel?.translatesAutoresizingMaskIntoConstraints = false
        self.dateTimeLabel?.numberOfLines = 1
        self.dateTimeLabel?.textColor = SendBirdUtils.UIColorFromRGB(0xacaab2)
        self.dateTimeLabel?.font = UIFont.systemFontOfSize(kFileLinkDateTimeFontSize)
        self.dateTimeLabel?.text = "11:24 PM"
        self.addSubview(self.dateTimeLabel!)
        
        self.fileImageView = UIImageView.init(image: UIImage.init(named: "_icon_file"))
        self.fileImageView?.translatesAutoresizingMaskIntoConstraints = false
        self.fileImageView?.contentMode = UIViewContentMode.ScaleAspectFit
        self.addSubview(self.fileImageView!)
        
        // Profile Image View
        self.addConstraint(NSLayoutConstraint.init(item: self.profileImageView!, attribute: NSLayoutAttribute.Bottom, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Bottom, multiplier: 1, constant: -kFileLinkCellBottomMargin))
        self.addConstraint(NSLayoutConstraint.init(item: self.profileImageView!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Leading, multiplier: 1, constant: kFileLinkCellLeftMargin))
        self.addConstraint(NSLayoutConstraint.init(item: self.profileImageView!, attribute: NSLayoutAttribute.Width, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: kFileLinkProfileWidth))
        self.addConstraint(NSLayoutConstraint.init(item: self.profileImageView!, attribute: NSLayoutAttribute.Height, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: kFileLinkProfileHeight))
        
        // Nickname Label
        self.addConstraint(NSLayoutConstraint.init(item: self.nicknameLabel!, attribute: NSLayoutAttribute.Bottom, relatedBy: NSLayoutRelation.Equal, toItem: self.fileImageView!, attribute: NSLayoutAttribute.Top, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.nicknameLabel!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Leading, multiplier: 1, constant: kFileLinkBalloonLeftPadding))
        self.addConstraint(NSLayoutConstraint.init(item: self.nicknameLabel!, attribute: NSLayoutAttribute.Width, relatedBy: NSLayoutRelation.LessThanOrEqual, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: kFileLinkWidth))
        
        // File Image View
        self.addConstraint(NSLayoutConstraint.init(item: self.fileImageView!, attribute: NSLayoutAttribute.Bottom, relatedBy: NSLayoutRelation.Equal, toItem: self.messageBackgroundImageView!, attribute: NSLayoutAttribute.Bottom, multiplier: 1, constant: -kFileLinkBalloonBottomPadding))
        self.addConstraint(NSLayoutConstraint.init(item: self.fileImageView!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Leading, multiplier: 1, constant: kFileLinkBalloonLeftPadding))
        self.addConstraint(NSLayoutConstraint.init(item: self.fileImageView!, attribute: NSLayoutAttribute.Width, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: kFileLinkWidth))
        self.addConstraint(NSLayoutConstraint.init(item: self.fileImageView!, attribute: NSLayoutAttribute.Height, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: kFileLinkHeight))
        
        // Message Background Image View
        self.addConstraint(NSLayoutConstraint.init(item: self.messageBackgroundImageView!, attribute: NSLayoutAttribute.Bottom, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Bottom, multiplier: 1, constant: kFileLinkCellBottomMargin))
        self.addConstraint(NSLayoutConstraint.init(item: self.messageBackgroundImageView!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Leading, multiplier: 1, constant: kFileLinkBalloonLeftPadding - 16))
        self.addConstraint(NSLayoutConstraint.init(item: self.messageBackgroundImageView!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self.fileImageView!, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: kFileLinkBalloonRightPadding))
        self.addConstraint(NSLayoutConstraint.init(item: self.messageBackgroundImageView!, attribute: NSLayoutAttribute.Top, relatedBy: NSLayoutRelation.Equal, toItem: self.nicknameLabel!, attribute: NSLayoutAttribute.Top, multiplier: 1, constant: -kFileLinkBalloonTopPadding))
        
        // DateTime Label
        self.addConstraint(NSLayoutConstraint.init(item: self.dateTimeLabel!, attribute: NSLayoutAttribute.Bottom, relatedBy: NSLayoutRelation.Equal, toItem: self.messageBackgroundImageView!, attribute: NSLayoutAttribute.Bottom, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.dateTimeLabel!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self.messageBackgroundImageView!, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: kFileLinkDateTimeLeftMarign))
    }
    
    func setModel(model: SendBirdFileLink) {
        self.fileLink = model
        let sender: SendBirdSender = (self.fileLink?.sender)!
        let ts: Int64 = (self.fileLink?.getMessageTimestamp())! / 1000
        self.dateTimeLabel?.text = SendBirdUtils.messageDateTime(NSTimeInterval(ts))
        self.nicknameLabel?.text = sender.name
        
        SendBirdUtils.loadImage(sender.imageUrl, imageView: self.profileImageView!, width: kFileLinkProfileWidth, height: kFileLinkProfileWidth)
        
        if self.fileLink?.fileInfo.type.hasPrefix("image") == true {
            SendBirdUtils.loadImage(model.fileInfo.url, imageView: self.fileImageView!, width: kFileLinkWidth, height: kFileLinkHeight)
        }
    }
    
    func getHeightOfViewCell(totalWidth: CGFloat) -> CGFloat {
        let nickname: String = self.fileLink!.sender.name
        var nicknameRect: CGRect?
        let nicknameAttribute = [NSFontAttributeName: UIFont.systemFontOfSize(12.0)]
        let attributedNickname: NSAttributedString = NSAttributedString.init(string: nickname, attributes: nicknameAttribute)
        
        nicknameRect = attributedNickname.boundingRectWithSize(CGSizeMake(kFileLinkWidth, CGFloat.max), options: NSStringDrawingOptions.UsesLineFragmentOrigin, context: nil)
        
        return nicknameRect!.size.height + kFileLinkCellBottomMargin + kFileLinkBalloonBottomPadding + kFileLinkHeight + kFileLinkBalloonTopPadding + kFileLinkCellTopMargin;
    }
}
